### PR TITLE
[1114] update app to not be API-only & dsi integration

### PR DESCRIPTION
**Pull request by misaka**
### Context

Splitting out the GUI and DSI stuff from the Get Information About Schools (GIAS) spike.

The GIAS spike is an exercise to see how information from GIAS matches to data on providers in TTAPI, for the purpose of importing URN and UKPRN data (the original Trello ticket appears to have been miss-placed). This was originally done as a local-only spike, but it seems useful to actually have in production sitting behind a feature-flag, and it also aligns with another goal to add an administrative console to TTAPI so we have to do less API work. Since these are effectively two separate requirements, the work to add the GUI and DSI integration has been split off into a separate PR (this one), and the GIAS spike will be a different one (#1825).

As an aside, this may also be used, eventually, as a starting point for moving Publish into TTAPI.

### Changes proposed in this pull request

Just add DSI sign-in integration with the necessary GUI bits (webpacker, govuk styling, etc). This is extracted from the GIAS spike which lives as a branch rebased ontop of this (see #1825).

### Guidance to review

Run locally, being sure to get and set the `dfe_signin.secret` setting (see @defong). Check that the styling looks good (have you run `yarn`? is webpacker compiling?) and that you can sign-in with DSI.

1. Run some commands
```bash
git checkout 452-add-dsi-integration
bundle install
yarn install
bundle exec rails server
```
2. Setup DSI creds
3. Visit https://localhost:3001/sign-in
4. login

You should then be redirected to the default Rails root page. You can double-check you're properly logged-in by revisiting the sign-in page (https://localhost:3001/sign-in).

### Checklist

- [x] Make sure all information from the Trello card is in here
- [x] Attach to Trello card
- [x] Rebased master
- [x] Cleaned commit history
- [x] Tested by running locally

### CODE REVIEW

**Comment by defong** [comment]
[[M] Add DSI integration to TTAPI (on the GIAS data branch)](https://trello.com/c/aIlnLcji/1114-m-add-dsi-integration-to-ttapi-on-the-gias-data-branch)

**Comment by misaka** [comment]
Looking really good. Should we remove the [452] labels on the commits? Probably don't need them.

**Comment by misaka** [comment]
Also, `452-add-gias-spike` rebases ontop with only 2 minor changes! 🎉

File: app/helpers/view_helper.rb [hunk 1]
```diff
@@ -0,0 +1,36 @@
+module ViewHelper
+  def govuk_link_to(body, url = body, html_options = { class: "govuk-link" })
```
Comment (gpeng) [inline app/helpers/view_helper.rb:2]: This is included from `govuk_components` so I don't think we need to define it. Same for `govuk_back_link_to`

File: app/mailers/application_mailer.rb [hunk 2]
```diff
@@ -0,0 +1,4 @@
+class ApplicationMailer < ActionMailer::Base
```
Comment (gpeng) [inline app/mailers/application_mailer.rb:1]: I don't think this is used. Do we need it?

File: app/models/dfe_sign_in_session.rb [hunk 3]
```diff
@@ -0,0 +1,69 @@
+# frozen_string_literal: true
+
+class DfESignInSession
+  # Let's try calling this UserSession instead of DfESignInUser ?
```
Comment (gpeng) [inline app/models/dfe_sign_in_session.rb:4]: 🤔

File: app/models/dfe_sign_in_session.rb [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+    }
+  end
+
+  def self.load_from_session(session)
+    dfe_sign_in_session = session["sign_in_session"]
+    return unless dfe_sign_in_session
+
+    # Users who signed in before session expiry was implemented will not have
+    # `last_active_at` set. In that case, force them to sign in again.
+    return unless dfe_sign_in_session["last_active_at"]
```
Comment (gpeng) [inline app/models/dfe_sign_in_session.rb:35]: I don't think we need this as all our signin users will have `last_active_at`

File: app/services/dfe_sign_in_users/update.rb [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+      }
+
+      attributes[:first_name] = dfe_sign_in_user.first_name if dfe_sign_in_user.first_name.present?
+      attributes[:last_name] = dfe_sign_in_user.last_name if dfe_sign_in_user.last_name.present?
+
+      @user.assign_attributes(attributes)
+    end
+
+    def call
+      @successful = user.valid? && user.save!
```
Comment (gpeng) [inline app/services/dfe_sign_in_users/update.rb:27]:
```suggestion
      @successful = user.save
```

File: app/views/layouts/_beta-banner.html.erb [hunk 6]
```diff
@@ -0,0 +1,8 @@
+<div class="govuk-phase-banner">
+  <p class="govuk-phase-banner__content">
+    <strong class="govuk-tag govuk-phase-banner__content__tag <%= beta_tag_environment_class %>">
+      <%= beta_banner_environment_label %>
+    </strong>
+    <span class="govuk-phase-banner__text">This is a new service – your <%= bat_contact_mail_to "feedback", subject: "Publish teacher training courses feedback" %> will help us to improve it.</span>
```
Comment (gpeng) [inline app/views/layouts/_beta-banner.html.erb:6]: I think we can probably remove the feedback link. Even if we don't this is the wrong text.

File: app/views/layouts/_flash.html.erb [hunk 7]
```diff
@@ -0,0 +1,31 @@
+<div class="grid-row">
```
Comment (gpeng) [inline app/views/layouts/_flash.html.erb:1]: Are we using flash messages yet? If not we do a lot less with them as we're using the form builder errors. I'd rather get rid of this and add a more up to date implementation when we need it.

File: app/views/layouts/mailer.html.erb [hunk 8]
```diff
@@ -0,0 +1,13 @@
+<!DOCTYPE html>
```
Comment (gpeng) [inline app/views/layouts/mailer.html.erb:1]: I don't think we're using this.

File: app/views/layouts/mailer.text.erb [hunk 9]
```diff
@@ -0,0 +1 @@
+<%= yield %>
```
Comment (gpeng) [inline app/views/layouts/mailer.text.erb:1]: ...or this

File: app/webpacker/packs/application.js [hunk 10]
```diff
@@ -0,0 +1,36 @@
+require.context("govuk-frontend/govuk/assets");
+
+import {initAutocomplete} from "../scripts/autocomplete";
```
Comment (gpeng) [inline app/webpacker/packs/application.js:3]: Do we need any of this?

File: app/webpacker/stylesheets/_notice-summary.scss [hunk 11]
```diff
@@ -0,0 +1,38 @@
+.app-notice-summary {
```
Comment (gpeng) [inline app/webpacker/stylesheets/_notice-summary.scss:1]: We can remove this and the other *-summary.scss if we don't include the flash partial.

File: app/webpacker/stylesheets/application.scss [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+$govuk-image-url-function: frontend-image-url;
+$success-green: #00823b;
+
+@import "~govuk-frontend/govuk/all";
+
+@import "admin-only";
+@import "banner";
+@import "environments";
+@import "header";
+@import "notice-summary";
```
Comment (gpeng) [inline app/webpacker/stylesheets/application.scss:19]: I don't think we need any of these from 19 down

File: app/webpacker/stylesheets/patterns/_pagination.scss [hunk 13]
```diff
@@ -0,0 +1,100 @@
+// TODO rewrite this to use more of GOVUK Frontend  mixin and styling
```
Comment (gpeng) [inline app/webpacker/stylesheets/patterns/_pagination.scss:1]: Or this... we have nothing to paginate yet and we have components for this stuff now.

File: app/webpacker/stylesheets/patterns/_tabs.scss [hunk 14]
```diff
@@ -0,0 +1,4 @@
+.app-tabs__list .govuk-tabs__list-item--selected {
```
Comment (gpeng) [inline app/webpacker/stylesheets/patterns/_tabs.scss:1]: This PR doesn't have any tabs and we have styles from the design system

File: spec/helpers/view_helper_spec.rb [hunk 15]
```diff
@@ -0,0 +1,53 @@
+require "rails_helper"
+
+feature "View helpers", type: :helper do
+  describe "#govuk_link_to" do
```
Comment (gpeng) [inline spec/helpers/view_helper_spec.rb:4]: These are in govuk_components so we don't need to test.

**Review by gpeng — CHANGES_REQUESTED** [review]
This looks good but I think there is a lot of stuff in here that we're not using yet and may well have to remove because we won't be following the publish patterns. e.g. we're using components or the form builder etc

File: app/services/dfe_sign_in_users/update.rb (hunk 5 above)
Comment (misaka) [reply app/services/dfe_sign_in_users/update.rb:27]: `user.save!` maybe?

File: app/webpacker/stylesheets/patterns/_pagination.scss (hunk 13 above)
Comment (misaka) [reply app/webpacker/stylesheets/patterns/_pagination.scss:1]: I worry this'll break the GIAS stuff, but ya ... that's not a good enough reason to keep it.

**Review by TobyRet — CHANGES_REQUESTED** [review]
Could we have some more context added to the PR description please. I don't understand some of the acronyms and it might be useful describe in more detail why these changes are being made. The Trello ticket does not give much context. 

Other than that, looking good. Agree with all the comments about fat trimming

File: app/services/dfe_sign_in_users/update.rb (hunk 5 above)
Comment (misaka) [reply app/services/dfe_sign_in_users/update.rb:27]: Ya, I think we should use `user.save!` and remove the `successful` memoisation and predicate.

I was looking at the same version of this in Register and this seems to be a pattern picked up from services used to update candidate details there. I'm not sure that makes sense for the login journey, an error should probably raise. The `successful?` predicate on `DFESignInUsers` certainly isn't used anywhere.

**Comment by misaka** [comment]
> Could we have some more context added to the PR description please. I don't understand some of the acronyms and it might be useful describe in more detail why these changes are being made. The Trello ticket does not give much context.

Context added @TobyRet, let me know if it's sufficient.

**Review by TobyRet — APPROVED** [review]
Thanks for updating the PR description @misaka 
LGTM once all the comments have been addressed

File: app/views/layouts/_beta-banner.html.erb (hunk 6 above)
Comment (defong) [reply app/views/layouts/_beta-banner.html.erb]: Removed.

File: app/views/layouts/_flash.html.erb (hunk 7 above)
Comment (defong) [reply app/views/layouts/_flash.html.erb]: Removed.

File: app/views/layouts/mailer.text.erb (hunk 9 above)
Comment (defong) [reply app/views/layouts/mailer.text.erb:1]: Do we need to remove from gov template as well as from register?

File: app/webpacker/packs/application.js (hunk 10 above)
Comment (defong) [reply app/webpacker/packs/application.js]: Removed.

File: app/webpacker/stylesheets/_notice-summary.scss (hunk 11 above)
Comment (defong) [reply app/webpacker/stylesheets/_notice-summary.scss]: Removed.

File: app/webpacker/stylesheets/application.scss (hunk 12 above)
Comment (defong) [reply app/webpacker/stylesheets/application.scss]: Removed.

File: app/webpacker/stylesheets/patterns/_pagination.scss (hunk 13 above)
Comment (defong) [reply app/webpacker/stylesheets/patterns/_pagination.scss]: Removed, beyond scope.

File: spec/helpers/view_helper_spec.rb (hunk 15 above)
Comment (defong) [reply spec/helpers/view_helper_spec.rb]: Removed.

File: app/helpers/view_helper.rb (hunk 1 above)
Comment (defong) [reply app/helpers/view_helper.rb]: Removed.

File: app/mailers/application_mailer.rb (hunk 2 above)
Comment (defong) [reply app/mailers/application_mailer.rb:1]: Do we need to remove from gov template as well as from register?

File: app/models/dfe_sign_in_session.rb (hunk 3 above)
Comment (defong) [reply app/models/dfe_sign_in_session.rb:4]: @misaka how many passing thought comments are added?

File: app/views/layouts/mailer.html.erb (hunk 8 above)
Comment (defong) [reply app/views/layouts/mailer.html.erb:1]: Do we need to remove from gov template as well as from register?

File: app/webpacker/stylesheets/patterns/_tabs.scss (hunk 14 above)
Comment (defong) [reply app/webpacker/stylesheets/patterns/_tabs.scss]: Removed.

File: package.json [hunk 16]
```diff
@@ -0,0 +1,26 @@
...
+  },
+  "dependencies": {
+    "@rails/webpacker": "^5.2.1",
+    "accessible-autocomplete": "^2.0.3",
+    "govuk-frontend": "^3.9.1",
+    "set-value": "^3.0.2"
+  },
+  "devDependencies": {
+    "jest": "^26.5.3",
+    "webpack-dev-server": "^3.11.0"
```
[the remark this answers is not in the document]
Comment (defong) [reply package.json:23]: Do we need to remove from gov template as well as from register?

**Comment by defong** [comment]
> This looks good but I think there is a lot of stuff in here that we're not using yet and may well have to remove because we won't be following the publish patterns. e.g. we're using components or the form builder etc

A larger and more relevant finessing was applied.

Predominantly using `govuk template` and augmenting with `register`s components where appropriate.

File: Dockerfile [hunk 17]
```diff
@@ -36,13 +36,24 @@ ADD Gemfile.lock $APP_HOME/Gemfile.lock
...
 ADD . $APP_HOME/
 
 COPY --from=middleman /public/ $APP_HOME/public/
 
 ENV COMMIT_SHA=${COMMIT_SHA}
 
+RUN ls /app/public/ && \ 
+    bundle exec rake assets:precompile && \
+    ls /app/public/packs/media/images
+    # rm -rf node_modules tmp
```
Comment (darokel) [inline Dockerfile:57]: can this be removed?

File: Gemfile [hunk 18]
```diff
@@ -6,6 +6,9 @@ ruby "2.7.2"
 # Bundle edge Rails instead: gem 'rails', github: 'rails/rails'
 gem "rails", "6.1.3"
 
+# Transpile app-like JavaScript. Read more: https://github.com/rails/webpacker
+gem "webpacker"
```
Comment (darokel) [inline Gemfile:10]: might be worth pinning this, the latest version 6 in beta introduces some breaking changes from the current stable release

File: app/views/layouts/application.html.erb [hunk 19]
```diff
@@ -0,0 +1,85 @@
...
+
+    <%= govuk_skip_link %>
+
+    <%= render Header::View.new(
+      service_name: I18n.t("service_name"),
+      items: header_items(@current_user),
+    ) %>
+
+    <%= render NavigationBar::View.new(
+      items: [
```
Comment (darokel) [inline app/views/layouts/application.html.erb:43]:
```suggestion
items: []
```

**Comment by darokel** [comment]
closing in favour of https://github.com/DFE-Digital/teacher-training-api/pull/1833

**Pull request closed without merging**